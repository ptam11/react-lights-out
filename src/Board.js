import React, { Component } from "react";
import Cell from "./Cell";
import './Board.css';
import uuid from 'uuid/v4';


/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - hasWon: boolean, true when board is all off
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

class Board extends Component {
  constructor(props) {
    super(props);
    this.state = {
      boxes: this._createBoard(),
      hasWon: false
    }
    this._createBoard = this._createBoard.bind(this);
    this._flipCellsAround = this._flipCellsAround.bind(this);
    this.flipCellsAndCheckForWin = this.flipCellsAndCheckForWin.bind(this);
    // TODO: set initial state
  }

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */

  static defaultProps = {
    nrows: 5,
    ncols: 5
  }

  _createBoard() {
    let board = [];
    // TODO: create array-of-arrays of true/false values
    const { nrows, ncols } = this.props;
    for (let y = 0; y < nrows; y++) {
      let row = Array.from({ length: ncols }, () => {
        let on = Math.random() < .50 ? true : false;
        return { isLit: on }
      });
      board.push(row)
    }

    return board
  }



  /** Internal method to copy a board and then flip the approiate cells **/

  _flipCellsAround(b, y, x) {
    let { ncols, nrows } = this.props;
    console.log({ b, y, x })
    // Make a copy of the board so that state is not changed outside of setState
    let board = b.map(r => [...r]);

    function flipCell(y, x) {
      // if this coord is actually on board, flip it
      if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
        board[y][x].isLit = !board[y][x].isLit;
      }
    }

    flipCell(y + 1, x)
    flipCell(y - 1, x)
    flipCell(y, x)
    flipCell(y, x + 1)
    flipCell(y, x - 1)

    // TODO: Flip all of the cells around the cell at y, x

    return board;
  }


  /** handle changing a cell: update board & determine if winner */
  flipCellsAndCheckForWin(y, x) {
    // console.log(this.state)
    let boxes = this._flipCellsAround(this.state.boxes, y, x);

    // win when every cell is turned off
    // TODO: determine is the game has been won
    let hasWon = this.state.boxes.every(row => {
      return row.every((box) => box.isLit === false)
    })

    this.setState({ boxes, hasWon });
  }


  /** Render game board or winning message. */

  render() {
    // let table = this.state.boxes.map(b => (
    //   <Cell isLit={b.isLit} flipCellsAroundMe={this._flipCellsAround}/>
    // ))
    const { ncols, nrows } = this.props;
    const { boxes } = this.state;
    let boxesArr = [];
    for (let y = 0; y < nrows; y++) {
      let arr = []
      for (let x = 0; x < ncols; x++) {
        let currBox = boxes[y][x];
        arr.push(<Cell x={x} y={y} key={uuid()} flipCellsAroundMe={this.flipCellsAndCheckForWin} isLit={currBox.isLit} />)
      }
      boxesArr.push(arr)
    }


    // if the game is won, just show a winning msg & render nothing else

    // TODO
    // make table board
    return this.state.hasWon ? <div>"You have won" </div>: (
      <table className="Board">
        <tbody>
          {boxesArr.map(row => {
            return (
              <tr key={uuid()}>
                {row}
              </tr>
            )
          })}
        </tbody>
      </table>)
  }
  // TODO
}


export default Board;
