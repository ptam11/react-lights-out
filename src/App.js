import React, {Component} from 'react';
import Board from './Board';
import './App.css';

/** Simple app that just shows the LightsOut game. */

class App extends Component {
  render() {
    return (
        <div className="App">
          <Board ncols={1} nrows={1}/>
        </div>
    );
  }
}

export default App;
