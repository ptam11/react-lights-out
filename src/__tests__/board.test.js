import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import React from 'react';
import ReactDOM from 'react-dom';
import App from '../App';
import Board from '../Board';

configure({ adapter: new Adapter() });

it('App renders without crashing', function() {
	shallow(<App />);
});

it('Board Component renders without errors', function() {
	const div = document.createElement('div');
	ReactDOM.render(<Board />, div);
	ReactDOM.unmountComponentAtNode(div);
});

it('matches snapshot', function() {
  let wrapper = shallow(<Board />);
  
	let serialized = toJson(wrapper);
	expect(serialized).toMatchSnapshot();
});

it('it should render starter board', function() {
	let wrapper = mount(<Board />);
	expect(wrapper.html()).toContain('<td class="Cell"></td>');
});

// it('it should ensure clicking will flip the right cells', function() {
//   let wrapper = mount(<Board />);
//   wrapper.setProps({ ncols= })
// 	wrapper.find("button[value='z']").simulate('click', { target: { value: 'z' } });

// 	expect(wrapper.html()).toContain(img1);
// });

// TODO https://airbnb.io/enzyme/docs/api/ReactWrapper/at.html 
// after finding similar elements do `at(index)`

it('it should return winning message', function() {
  let wrapper = mount(<Board />);
  wrapper.setProps({ ncols: 2, nrows: 2 }) 
  wrapper.setState({ 
    hasWon: true, 
    boxes: [
      [{isLit: false}, {isLit: false}],
      [{isLit: false}, {isLit: false}]
    ]
  })
  console.log(wrapper.debug())
	expect(wrapper.html()).toContain("You have won");
});
