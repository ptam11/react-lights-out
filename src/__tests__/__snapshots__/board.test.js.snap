// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<table
  className="Board"
>
  <tbody>
    <tr
      key="d1f9ee7c-9e9d-4efe-af6b-ef3857536dec"
    >
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={false}
        key="1acb3308-455b-4bca-a749-e12cd8e766dd"
        x={0}
        y={0}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={true}
        key="2e7dad1b-ef86-42fb-b277-4e176ca8c096"
        x={1}
        y={0}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={false}
        key="5e3d3cf0-1694-44fe-9fcd-83ca87fd2254"
        x={2}
        y={0}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={false}
        key="83951ad7-bdd5-45ad-97ab-9dec031a7699"
        x={3}
        y={0}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={true}
        key="96453a3e-76cc-4381-af0e-ae021c0d8f53"
        x={4}
        y={0}
      />
    </tr>
    <tr
      key="62abe792-cfd3-4034-a5c7-b72e1dbd5a9e"
    >
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={false}
        key="f4db217f-3f51-4444-9b74-2f1fbe89ae22"
        x={0}
        y={1}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={false}
        key="6b9f8e33-198f-45e7-a1ce-e30455973801"
        x={1}
        y={1}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={true}
        key="f1e47e38-183f-4c3c-a861-12382cdfa9c7"
        x={2}
        y={1}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={false}
        key="f18e15b8-d70e-40c8-b0b3-07f5fe38ee7e"
        x={3}
        y={1}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={true}
        key="aa1843d2-c9e6-462c-a70b-d0aeccfacb47"
        x={4}
        y={1}
      />
    </tr>
    <tr
      key="2920ac43-6d15-496c-ab18-d46018e8760b"
    >
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={false}
        key="8242bfa9-0573-42a8-be71-d7c9ca989044"
        x={0}
        y={2}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={true}
        key="a85af795-c954-46e0-adbb-a66e42808ba5"
        x={1}
        y={2}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={true}
        key="ffd1439b-35df-499a-9dcf-2a53bb316d3e"
        x={2}
        y={2}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={false}
        key="39690f91-166f-43b4-a507-b419eeaa9254"
        x={3}
        y={2}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={true}
        key="16f50193-6cd1-4267-8ac8-a7fb81b433de"
        x={4}
        y={2}
      />
    </tr>
    <tr
      key="ee5b9903-9843-4577-bc01-7824a329cf6c"
    >
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={true}
        key="27db04eb-ff7a-4b78-8312-e0fc1d82e110"
        x={0}
        y={3}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={true}
        key="0c69ea62-f9ba-404e-94d7-fd3165a1125b"
        x={1}
        y={3}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={false}
        key="f1ff9315-c0f4-489e-b4b2-c3a394cd51dc"
        x={2}
        y={3}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={true}
        key="184f59c7-f53a-421a-aa16-8e9b26d1d00f"
        x={3}
        y={3}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={false}
        key="42c3af85-117d-4a1c-ae73-45d4563f3c70"
        x={4}
        y={3}
      />
    </tr>
    <tr
      key="25ddecab-d975-49f2-afa0-e0b6e254f1a0"
    >
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={false}
        key="7ee801b8-445e-4291-838a-1a99cdae1dca"
        x={0}
        y={4}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={false}
        key="7da8387e-4501-44c8-a09a-2005372f50e1"
        x={1}
        y={4}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={true}
        key="87014286-6ee7-44ee-824f-fc081f016ccd"
        x={2}
        y={4}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={true}
        key="76140c1d-ae12-4df3-9e8c-855513fbc43a"
        x={3}
        y={4}
      />
      <Cell
        flipCellsAroundMe={[Function]}
        isLit={true}
        key="a1cb1790-d16e-4801-a532-be0f2ecb168b"
        x={4}
        y={4}
      />
    </tr>
  </tbody>
</table>
`;
